Sync Stalling

### Description
Three times in a row since updating this afternoon my Medalla sync has stalled.  At INFO level there's no sign of anything going wrong, it just stops progressing after an hour or two.  Still has peers, but seems to have stopped requesting new blocks entirely.  I tried again with debug level enabled for the `tech.pegasys.teku.sync` package and got the exception below:

```
 slot 435936 by replaying 20 blocks. Took 2252ms","throwable":""}
{"timestamp":"2020-10-20T06:36:48,547","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552483, Head slot: 435980, Connected peers: 74","throwable":""}
{"timestamp":"2020-10-20T06:36:56,389","level":"DEBUG","thread":"forkchoice-0","class":"PeerSync","message":"Request 200 blocks starting at 436012 from peer 16Uiu2HAkyjDgW6orCnb1RLkm21Voho617U8ay89tZGzUQcx8fbjL","throwable
":""}
{"timestamp":"2020-10-20T06:36:56,806","level":"WARN","thread":"nioEventLoopGroup-3-2","class":"DefaultChannelPipeline","message":"An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usuall
y means the last handler in the pipeline did not handle the exception.","throwable":" io.netty.channel.ChannelPipelineException: tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.handlerRemoved() has thrown
 an exception.\n\tat io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:640)\n\tat io.netty.channel.DefaultChannelPipeline.destroyDown(DefaultChannelPipeline.java:876)\n\tat io.netty.ch
annel.DefaultChannelPipeline.destroyUp(DefaultChannelPipeline.java:844)\n\tat io.netty.channel.DefaultChannelPipeline.destroy(DefaultChannelPipeline.java:836)\n\tat io.netty.channel.DefaultChannelPipeline.access$700(Defaul
tChannelPipeline.java:46)\n\tat io.netty.channel.DefaultChannelPipeline$HeadContext.channelUnregistered(DefaultChannelPipeline.java:1392)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(Abstr
actChannelHandlerContext.java:198)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:184)\n\tat io.netty.channel.DefaultChannelPipeline.fireChannelUnregistere
d(DefaultChannelPipeline.java:821)\n\tat io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:826)\n\tat io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)\n\ta
t io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)\n\tat io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)\n\tat io.netty.util.concurrent.SingleThreadEventExecut
or$4.run(SingleThreadEventExecutor.java:989)\n\tat io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat
 java.base/java.lang.Thread.run(Thread.java:832)\nCaused by: java.lang.IllegalStateException: Trying to reuse disposable LengthPrefixedPayloadDecoder\n\tat tech.pegasys.teku.networking.eth2.rpc.core.encodings.LengthPrefixe
dPayloadDecoder.complete(LengthPrefixedPayloadDecoder.java:108)\n\tat tech.pegasys.teku.networking.eth2.rpc.core.RpcResponseDecoder.complete(RpcResponseDecoder.java:115)\n\tat tech.pegasys.teku.networking.eth2.rpc.core.Rpc
ResponseDecoder.close(RpcResponseDecoder.java:105)\n\tat tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.abortRequest(Eth2OutgoingRequestHandler.java:222)\n\tat tech.pegasys.teku.networking.eth2.rpc.c
ore.Eth2OutgoingRequestHandler.abortRequest(Eth2OutgoingRequestHandler.java:212)\n\tat tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.readComplete(Eth2OutgoingRequestHandler.java:159)\n\tat tech.pega
sys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.onRemoteWriteClosed(RpcHandler.java:191)\n\tat tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.onChannelClosed(RpcHandler.java:197)\n\tat tech.pega
sys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.handlerRemoved(RpcHandler.java:178)\n\tat io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:946)\n\tat io.netty.ch
annel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)\n\t... 16 more\n"}
{"timestamp":"2020-10-20T06:36:57,148","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552484, Head slot: 436011, Connected peers: 73","throwable":""}
{"timestamp":"2020-10-20T06:37:09,112","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552485, Head slot: 436013, Connected peers: 74","throwable":""}
{"timestamp":"2020-10-20T06:37:21,170","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552486, Head slot: 436013, Connected peers: 73","throwable":""}
{"timestamp":"2020-10-20T06:37:33,123","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552487, Head slot: 436013, Connected peers: 74","throwable":""}
{"timestamp":"2020-10-20T06:37:45,108","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552488, Head slot: 436013, Connected peers: 74","throwable":""}
{"timestamp":"2020-10-20T06:37:57,105","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552489, Head slot: 436013, Connected peers: 74","throwable":""}
{"timestamp":"2020-10-20T06:38:09,365","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552490, Head slot: 436013, Connected peers: 73","throwable":""}
{"timestamp":"2020-10-20T06:38:21,105","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552491, Head slot: 436013, Connected peers: 73","throwable":""}
{"timestamp":"2020-10-20T06:38:22,958","level":"WARN","thread":"p2p-async-0","class":"LoggingWeakSubjectivityViolationPolicy","message":"The latest finalized checkpoint at epoch 13623 (67417 active validators) fell outside
 of the weak subjectivity period after epoch 14721, which is prior to the current epoch 17265.","throwable":""}
{"timestamp":"2020-10-20T06:38:22,958","level":"WARN","thread":"p2p-async-0","class":"teku-status-log","message":"The latest finalized checkpoint at epoch 13623 is outside of the weak subjectivity period.  Please supply a
recent weak subjectivity checkpoint using --ws-checkpoint=<BLOCK_ROOT>:<EPOCH>.","throwable":""}
{"timestamp":"2020-10-20T06:38:33,234","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552492, Head slot: 436013, Connected peers: 72","throwable":""}
{"timestamp":"2020-10-20T06:38:45,113","level":"INFO","thread":"TimeTickChannel-0","class":"teku-event-log","message":"Sync Event  *** Current slot: 552493, Head slot: 436013, Connected peers: 72","throwable":""}
```

Formatted trace:
```
An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.","throwable":" io.netty.channel.ChannelPipelineException: tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.handlerRemoved() has thrown an exception.
	at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:640)
	at io.netty.channel.DefaultChannelPipeline.destroyDown(DefaultChannelPipeline.java:876)
	at io.netty.channel.DefaultChannelPipeline.destroyUp(DefaultChannelPipeline.java:844)
	at io.netty.channel.DefaultChannelPipeline.destroy(DefaultChannelPipeline.java:836)
	at io.netty.channel.DefaultChannelPipeline.access$700(DefaultChannelPipeline.java:46)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelUnregistered(DefaultChannelPipeline.java:1392)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:198)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:184)
	at io.netty.channel.DefaultChannelPipeline.fireChannelUnregistered(DefaultChannelPipeline.java:821)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:826)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.IllegalStateException: Trying to reuse disposable LengthPrefixedPayloadDecoder
	at tech.pegasys.teku.networking.eth2.rpc.core.encodings.LengthPrefixedPayloadDecoder.complete(LengthPrefixedPayloadDecoder.java:108)
	at tech.pegasys.teku.networking.eth2.rpc.core.RpcResponseDecoder.complete(RpcResponseDecoder.java:115)
	at tech.pegasys.teku.networking.eth2.rpc.core.RpcResponseDecoder.close(RpcResponseDecoder.java:105)
	at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.abortRequest(Eth2OutgoingRequestHandler.java:222)
	at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.abortRequest(Eth2OutgoingRequestHandler.java:212)
	at tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.readComplete(Eth2OutgoingRequestHandler.java:159)
	at tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.onRemoteWriteClosed(RpcHandler.java:191)
	at tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.onChannelClosed(RpcHandler.java:197)
	at tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.handlerRemoved(RpcHandler.java:178)
	at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:946)
	at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
	... 16 more
```

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/Eth2OutgoingRequestHandler.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/Eth2OutgoingRequestHandler.java
@@ -218,12 +218,15 @@ public class Eth2OutgoingRequestHandler<TRequest extends RpcRequest, TResponse>
       return;
     }
 
-    // releasing any resources
-    responseDecoder.close();
-
     LOG.trace("Abort request: {}", error.getMessage());
-    rpcStream.closeAbruptly().reportExceptions();
-    responseProcessor.finishProcessing().always(() -> responseStream.completeWithError(error));
+
+    // releasing any resources
+    try {
+      responseDecoder.close();
+      rpcStream.closeAbruptly().reportExceptions();
+    } finally {
+      responseProcessor.finishProcessing().always(() -> responseStream.completeWithError(error));
+    }
   }
 
   private void ensureFirstBytesArriveWithinTimeLimit(final RpcStream stream) {

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/RpcResponseDecoder.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/RpcResponseDecoder.java
@@ -25,6 +25,7 @@ import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import tech.pegasys.teku.networking.eth2.rpc.core.RpcException.PayloadTruncatedException;
 import tech.pegasys.teku.networking.eth2.rpc.core.RpcException.RpcErrorMessage;
+import tech.pegasys.teku.networking.eth2.rpc.core.encodings.ByteBufDecoder;
 import tech.pegasys.teku.networking.eth2.rpc.core.encodings.RpcByteBufDecoder;
 import tech.pegasys.teku.networking.eth2.rpc.core.encodings.RpcEncoding;
 
@@ -101,12 +102,8 @@ public class RpcResponseDecoder<T> {
   }
 
   public void close() {
-    try {
-      complete();
-    } catch (RpcException e) {
-      // decoders should release any resources despite throwing exception
-      LOG.trace("Ignoring any complete() exceptions when close(): %s", e);
-    }
+    payloadDecoder.ifPresent(ByteBufDecoder::close);
+    errorDecoder.ifPresent(ByteBufDecoder::close);
   }
 
   public void complete() throws RpcException {

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/RpcResponseDecoder.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/RpcResponseDecoder.java
@@ -21,8 +21,6 @@ import io.netty.buffer.ByteBuf;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
 import tech.pegasys.teku.networking.eth2.rpc.core.RpcException.PayloadTruncatedException;
 import tech.pegasys.teku.networking.eth2.rpc.core.RpcException.RpcErrorMessage;
 import tech.pegasys.teku.networking.eth2.rpc.core.encodings.ByteBufDecoder;
@@ -35,8 +33,6 @@ import tech.pegasys.teku.networking.eth2.rpc.core.encodings.RpcEncoding;
  * @param <T>
  */
 public class RpcResponseDecoder<T> {
-  private static final Logger LOG = LogManager.getLogger();
-
   private Optional<Integer> respCodeMaybe = Optional.empty();
   private Optional<RpcByteBufDecoder<T>> payloadDecoder = Optional.empty();
   private Optional<RpcByteBufDecoder<RpcErrorMessage>> errorDecoder = Optional.empty();

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/AbstractByteBufDecoder.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/AbstractByteBufDecoder.java
@@ -65,9 +65,19 @@ public abstract class AbstractByteBufDecoder<TMessage, TException extends Except
 
   @Override
   public void complete() throws TException {
+    try {
+      if (compositeByteBuf.isReadable()) {
+        throwUnprocessedDataException(compositeByteBuf.readableBytes());
+      }
+    } finally {
+      close();
+    }
+  }
+
+  @Override
+  public void close() {
     if (compositeByteBuf.isReadable()) {
       compositeByteBuf.release();
-      throwUnprocessedDataException(compositeByteBuf.readableBytes());
     }
   }
 

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/ByteBufDecoder.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/ByteBufDecoder.java
@@ -53,10 +53,22 @@ public interface ByteBufDecoder<TMessage, TException extends Exception> {
 
   /**
    * Tells the decoder that the input stream is over and no more bytes would be supplied. After
-   * {@code complete()} call calling any Decoder method would result in error If any buffered data
-   * still remains unprocessed in the decoder it is released and {@link TException} is thrown
+   * {@code complete()} call calling any Decoder method would result in error
+   *
+   * The {@link #close()} should be invoked by implementation on either success or error so
+   * on {@code complete()} return all resources should be released by a decoder
+   *
+   * To abort the decoder use {@link #close()} method
    *
    * @throws TException if any unprocessed data left
    */
   void complete() throws TException;
+
+  /**
+   * Releases all associated resources. Normally should be used to release resources on abrupt
+   * completion. Use {@link #complete()} method for regular decoder completion.
+   *
+   * After {@code close()} calling any Decoder method would result in error.
+   */
+  void close();
 }

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/ByteBufDecoder.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/ByteBufDecoder.java
@@ -55,10 +55,10 @@ public interface ByteBufDecoder<TMessage, TException extends Exception> {
    * Tells the decoder that the input stream is over and no more bytes would be supplied. After
    * {@code complete()} call calling any Decoder method would result in error
    *
-   * The {@link #close()} should be invoked by implementation on either success or error so
-   * on {@code complete()} return all resources should be released by a decoder
+   * <p>The {@link #close()} should be invoked by implementation on either success or error so on
+   * {@code complete()} return all resources should be released by a decoder
    *
-   * To abort the decoder use {@link #close()} method
+   * <p>To abort the decoder use {@link #close()} method
    *
    * @throws TException if any unprocessed data left
    */
@@ -68,7 +68,7 @@ public interface ByteBufDecoder<TMessage, TException extends Exception> {
    * Releases all associated resources. Normally should be used to release resources on abrupt
    * completion. Use {@link #complete()} method for regular decoder completion.
    *
-   * After {@code close()} calling any Decoder method would result in error.
+   * <p>After {@code close()} calling any Decoder method would result in error.
    */
   void close();
 }

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/LengthPrefixedEncoding.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/LengthPrefixedEncoding.java
@@ -33,6 +33,9 @@ public class LengthPrefixedEncoding implements RpcEncoding {
 
         @Override
         public void complete() {}
+
+        @Override
+        public void close() {}
       };
 
   private final String name;

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/LengthPrefixedPayloadDecoder.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/LengthPrefixedPayloadDecoder.java
@@ -108,32 +108,28 @@ class LengthPrefixedPayloadDecoder<T> implements RpcByteBufDecoder<T> {
       throw new IllegalStateException("Trying to reuse disposable LengthPrefixedPayloadDecoder");
     }
     disposed = true;
-    RpcException err = null;
-    if (varIntDecoder.isPresent()) {
-      try {
-        varIntDecoder.get().complete();
-      } catch (Exception e) {
-        // ignore any exception, call complete() just to release resources
-      }
-      // if varIntDecoder exists then payload length was not read completely
-      err = new MessageTruncatedException();
-    }
-    if (decompressor.isPresent()) {
-      try {
-        decompressor.get().complete();
-      } catch (CompressionException e) {
-        // ignore any exception, call complete() just to release resources
-      }
-      // if decompressor still exists then not enough data was fed to it
-      err = new PayloadTruncatedException();
-    }
-    if (!decoded && err == null) {
-      err = new MessageTruncatedException();
-    }
 
-    if (err != null) {
-      throw err;
+    try {
+      if (varIntDecoder.isPresent()) {
+        // if varIntDecoder exists then payload length was not read completely
+        throw new MessageTruncatedException();
+      }
+      if (decompressor.isPresent()) {
+        // if decompressor still exists then not enough data was fed to it
+        throw new PayloadTruncatedException();
+      }
+      if (!decoded) {
+        throw new MessageTruncatedException();
+      }
+    } finally {
+      close();
     }
+  }
+
+  @Override
+  public void close() {
+    varIntDecoder.ifPresent(AbstractByteBufDecoder::close);
+    decompressor.ifPresent(ByteBufDecoder::close);
   }
 
   /** Decode the length-prefix header, which contains the length of the uncompressed payload */

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/noop/NoopCompressor.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/noop/NoopCompressor.java
@@ -47,6 +47,11 @@ public class NoopCompressor implements Compressor {
       decoder.complete();
       disposed = true;
     }
+
+    @Override
+    public void close() {
+      decoder.close();
+    }
   }
 
   @Override

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/snappy/SnappyFramedCompressor.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/snappy/SnappyFramedCompressor.java
@@ -88,13 +88,7 @@ public class SnappyFramedCompressor implements Compressor {
           }
         } catch (Exception e) {
           broken = true;
-          try {
-            snappyFrameDecoder.complete();
-          } catch (Exception ex) {
-            LOG.trace("Failed to complete snappy frame decoding", ex);
-          }
-          decodedSnappyFrames.forEach(ReferenceCounted::release);
-          decodedSnappyFrames.clear();
+          close();
           throw e;
         }
       }
@@ -103,16 +97,24 @@ public class SnappyFramedCompressor implements Compressor {
 
     @Override
     public void complete() throws CompressionException {
-      if (broken) throw new CompressionException("Compressed stream is broken");
-      if (disposed) throw new DisposedDecompressorException();
-      disposed = true;
-      boolean unreturnedFrames = !decodedSnappyFrames.isEmpty();
+      try {
+        if (broken) throw new CompressionException("Compressed stream is broken");
+        if (disposed) throw new DisposedDecompressorException();
+        disposed = true;
+        boolean unreturnedFrames = !decodedSnappyFrames.isEmpty();
+        if (unreturnedFrames) {
+          throw new PayloadSmallerThanExpectedException("Unread uncompressed frames on complete");
+        }
+      } finally {
+        close();
+      }
+    }
+
+    @Override
+    public void close() {
       decodedSnappyFrames.forEach(ReferenceCounted::release);
       decodedSnappyFrames.clear();
-      snappyFrameDecoder.complete();
-      if (unreturnedFrames) {
-        throw new PayloadSmallerThanExpectedException("Unread uncompressed frames on complete");
-      }
+      snappyFrameDecoder.close();
     }
   }
 

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/snappy/SnappyFramedCompressor.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/snappy/SnappyFramedCompressor.java
@@ -96,10 +96,10 @@ public class SnappyFramedCompressor implements Compressor {
     public void complete() throws CompressionException {
       try {
         if (broken) {
-            throw new CompressionException("Compressed stream is broken");
+          throw new CompressionException("Compressed stream is broken");
         }
         if (disposed) {
-            throw new DisposedDecompressorException();
+          throw new DisposedDecompressorException();
         }
         disposed = true;
         boolean unreturnedFrames = !decodedSnappyFrames.isEmpty();

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/snappy/SnappyFramedCompressor.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/snappy/SnappyFramedCompressor.java
@@ -95,8 +95,12 @@ public class SnappyFramedCompressor implements Compressor {
     @Override
     public void complete() throws CompressionException {
       try {
-        if (broken) throw new CompressionException("Compressed stream is broken");
-        if (disposed) throw new DisposedDecompressorException();
+        if (broken) {
+            throw new CompressionException("Compressed stream is broken");
+        }
+        if (disposed) {
+            throw new DisposedDecompressorException();
+        }
         disposed = true;
         boolean unreturnedFrames = !decodedSnappyFrames.isEmpty();
         if (unreturnedFrames) {

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/snappy/SnappyFramedCompressor.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/encodings/compression/snappy/SnappyFramedCompressor.java
@@ -19,8 +19,6 @@ import io.netty.util.ReferenceCounted;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
 import org.apache.tuweni.bytes.Bytes;
 import tech.pegasys.teku.networking.eth2.rpc.core.encodings.compression.Compressor;
 import tech.pegasys.teku.networking.eth2.rpc.core.encodings.compression.exceptions.CompressionException;
@@ -30,7 +28,6 @@ import tech.pegasys.teku.networking.eth2.rpc.core.encodings.compression.exceptio
 
 /** Implements snappy compression using the "framed" / streaming format. */
 public class SnappyFramedCompressor implements Compressor {
-  private static final Logger LOG = LogManager.getLogger();
 
   private class SnappyFramedDecompressor implements Decompressor {
     private final SnappyFrameDecoder snappyFrameDecoder = new SnappyFrameDecoder();

--- a/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/core/Eth2OutgoingRequestHandlerTest.java
+++ b/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/core/Eth2OutgoingRequestHandlerTest.java
@@ -271,6 +271,21 @@ public abstract class Eth2OutgoingRequestHandlerTest
   }
 
   @Test
+  public void shouldCompleteExceptionallyWhenClosedWithTruncatedMessage() {
+    sendInitialPayload();
+
+    timeProvider.advanceTimeByMillis(100);
+    final Bytes chunkBytes = chunks.get(0);
+    deliverBytes(chunkBytes.slice(0, chunkBytes.size() - 1));
+
+    asyncRequestRunner.executeQueuedActions();
+
+    complete();
+
+    assertThat(finishedProcessingFuture).isCompletedExceptionally();
+  }
+
+  @Test
   public void doNotDisconnectsIfSecondChunkReceivedInTime() throws Exception {
     sendInitialPayload();
 

